lib/join_cluster.js

```diff
@@ -0,0 +1,402 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+'use strict';
+
+var captureHost = require('./util.js').captureHost;
+var errors = require('./errors.js');
+var isEmptyArray = require('./util.js').isEmptyArray;
+var numOrDefault = require('./util.js').numOrDefault;
+var safeParse = require('./util.js').safeParse;
+var TypedError = require('error/typed');
+
+var JoinAbortedError = TypedError({
+    type: 'ringpop.join-aborted',
+    message: 'Join aborted because `{reason}`',
+    reason: null
+});
+
+var JoinDurationExceededError = TypedError({
+    type: 'ringpop.join-duration-exceeded',
+    message: 'Join duration of `{duration}` exceeded max `{max}`',
+    duration: null,
+    max: null
+});
+
+var JOIN_SIZE = 3;
+var JOIN_TIMEOUT = 1000;
+var MAX_JOIN_DURATION = 120000;
```

Comment:
2 minutes is a bit long. if a join can't take place within a few seconds, there's usually something the matter. the reason for the longer time period is because it may give people a bit more time to look to see what's wrong without having to do a full restart of all instances. either way, 30s or 120s is rather arbitrary.